File: lintcode/longest-palindromic-substring.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/longest-palindromic-substring/

class Solution:
    # @param {string} s input string
    # @return {string} the longest palindromic substring
    def longestPalindrome(self, s):
        # Write your code here
        result = ''

        # For odd palindrome.
        for i in xrange(len(s)):
            j = 1
            while i-j >= 0 and i+j <= len(s)-1 and s[i-j] == s[i+j]:
                j += 1
            if len(result) < 2*j-1:
                result = s[i-j+1:i+j]

        # For even palindrome.
        for i in xrange(len(s)):
            j = 0
            while i-j >= 0 and i+j+1 <=len(s)-1 and s[i-j] == s[i+j+1]:
                j += 1
            if j == 0:
                if len(result) < 1:
                    result = s[i]
            else:
                if len(result) < 2*j:
                    result = s[i-j+1:i+j+1]

        return result

if __name__ == '__main__':
    s = Solution()
    print s.longestPalindrome('abcdzdcab')
    print s.longestPalindrome('abcddcab')
    print s.longestPalindrome('abcd')

File: misc/amazon/longest-palindromic-substring.cc
// Question: Longest palindromic substring.

#include <string>
#include <iostream>
using namespace std;

string longest_palindromic_substring(string s) {
    if (s.empty()) {
        return "";
    }

    string ans = s.substr(0, 1);

    for (int i = 0; i < s.size(); ++i) {
        int j = 0;
        while (i-j >= 0 && i-j < s.size() && s[i-j] == s[i+j]) {
            ++j;
        }
        if (2*j-1 > ans.size()) {
            ans = s.substr(i-j+1, 2*j-1);
        }
    }

    for (int i = 0; i < s.size(); ++i) {
        int j = 0;
        while (i-j >= 0 && i+j+1 < s.size() && s[i-j] == s[i+j+1]) {
            ++j;
        }
        if (2*j > ans.size()) {
            ans = s.substr(i-j+1, 2*j);
        }
    }

    return ans;
}

int main(int argc, char** argv) {
    cout << longest_palindromic_substring("abcdzdcab") << endl;
    cout << longest_palindromic_substring("abcddcab") << endl;
    cout << longest_palindromic_substring("abcd") << endl;
}
